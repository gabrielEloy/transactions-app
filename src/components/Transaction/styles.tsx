import {StyleSheet} from 'react-native';
import {COLORS} from 'src/themes/main';

const styles = StyleSheet.create({
  container: {
    backgroundColor: COLORS.contrastBackground,
    height: 75,
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 3.84,

    elevation: 5,
    flexDirection: 'row',
    borderRadius: 3,
  },
  indicatorContainer: {
    flex: 2,
    justifyContent: 'center',
    alignItems: 'center',
  },
  descriptionContainer: {
    flex: 4,
    justifyContent: 'center',
  },
  valueContainer: {
    flex: 2,
    alignItems: 'center',
    justifyContent: 'center',
  },
  debit: {
    color: COLORS.debit,
  },
  credit: {
    color: COLORS.credit,
  },
  dateText: {
    fontSize: 12,
    fontFamily: 'Exo-Medium',
    color: COLORS.gray,
  },

  descriptionText: {
    marginBottom: 5,
    fontSize: 20,
    fontFamily: 'Exo-Medium',
    color: COLORS.light,
  },
  gradientStyle: {
    width: 45,
    height: 45,
    borderRadius: 50,
  },
});

export default styles;
