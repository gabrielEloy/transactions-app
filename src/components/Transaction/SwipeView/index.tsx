import React from 'react';
import {TouchableOpacity, Text} from 'react-native';
import ITransactions from 'interfaces/ITransactions';
import styles from './styles';

interface Props {
  onPress: (transaction: ITransactions) => void;
}

const SwipeView = ({onPress}: Props) => (
  <TouchableOpacity
    onPress={onPress}
    style={styles.container}
    testID="deleteTransaction">
    <Text style={styles.deleteText}>Delete</Text>
  </TouchableOpacity>
);

export default SwipeView;
