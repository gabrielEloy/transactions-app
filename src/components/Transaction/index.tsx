import React from 'react';
import {View, Text} from 'react-native';
import moment from 'moment';
import Swipeable from 'react-native-gesture-handler/Swipeable';

import LinearGradient from 'react-native-linear-gradient';
import {getCurrencyString} from 'src/helpers/currency';
import IValue from 'src/interfaces/IValue';
import styles from './styles';
import ITransactions from '../../interfaces/ITransaction';
import SwipeView from './SwipeView';

interface Props {
  id: string;
  description: string;
  value: IValue;
  debit: boolean;
  date: string;
  style?: object;
  onSwipe: (transaction: ITransactions) => void;
}

const Transaction = ({
  debit,
  value,
  description,
  style,
  date,
  id,
  onSwipe,
}: Props) => {
  const [gradientColors, valueStyle] = debit
    ? [['#FFB4B5', '#FF454A'], styles.debit]
    : [['#BEFFA0', '#24C832'], styles.credit];

  const handleSwipe = () => {
    onSwipe({debit, value, description, id, date});
  };

  return (
    <Swipeable
      rightThreshold={100}
      overshootRight={false}
      renderRightActions={() => <SwipeView onPress={handleSwipe} />}>
      <View style={[styles.container, style]}>
        <View style={styles.indicatorContainer}>
          <LinearGradient
            colors={gradientColors}
            start={{x: 0.8, y: 0}}
            end={{x: 0.2, y: 0.7}}
            style={styles.gradientStyle}
          />
        </View>
        <View style={styles.descriptionContainer}>
          <Text adjustsFontSizeToFit style={styles.descriptionText}>
            {description}
          </Text>
          <Text
            testID="transactionDate"
            adjustsFontSizeToFit
            style={styles.dateText}>
            {moment(date).format('DD/MM/YYYY, h:mm:ss a')}
          </Text>
        </View>
        <View style={styles.valueContainer}>
          <Text
            testID="transactionValue"
            adjustsFontSizeToFit
            style={valueStyle}>
            {getCurrencyString(value)}
          </Text>
        </View>
      </View>
    </Swipeable>
  );
};

export default Transaction;
