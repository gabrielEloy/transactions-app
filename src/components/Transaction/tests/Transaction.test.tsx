import React from 'react';
import Transaction from 'src/components/Transaction';

import {render} from 'react-native-testing-library';

describe('Button', () => {
  const transaction = {
    date: '2020-08-17T19:53:14.283Z',
    debit: false,
    description: 'teste2',
    id: '2020-08-17T19:53:14.283Z-teste2',
    value: {currency: 'R$', value: 3500},
  };
  const onSwipe = jest.fn();
  const tree = render(<Transaction {...transaction} onSwipe={onSwipe} />);

  test('Should render correctly', () => {
    expect(tree).toMatchSnapshot();
  });

  test('Should correctly format date', async () => {
    const {queryByText} = await render(
      <Transaction {...transaction} onSwipe={onSwipe} />,
    );

    expect(queryByText('17/08/2020, 4:53:14 pm')).not.toBeNull();
  });

  test('Should correctly format currency', () => {
    const {queryByText} = render(
      <Transaction {...transaction} onSwipe={onSwipe} />,
    );

    expect(queryByText('R$ 3.500,00')).not.toBeNull();
  });
});
