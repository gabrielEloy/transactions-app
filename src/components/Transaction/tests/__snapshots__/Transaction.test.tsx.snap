// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Should render correctly 1`] = `
<View
  collapsable={false}
  forwardedRef={[Function]}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  onLayout={[Function]}
  style={
    Object {
      "overflow": "hidden",
    }
  }
>
  <View
    style={
      Object {
        "bottom": 0,
        "flexDirection": "row-reverse",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": Array [
          Object {
            "translateX": -10000,
          },
        ],
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "red",
          "borderBottomRightRadius": 5,
          "borderTopRightRadius": 5,
          "height": 73,
          "justifyContent": "center",
          "opacity": 1,
          "transform": Array [
            Object {
              "translateX": -10,
            },
            Object {
              "translateY": 6,
            },
          ],
          "width": 100,
        }
      }
      testID="deleteTransaction"
    >
      <Text
        style={
          Object {
            "color": "#FBFBFB",
          }
        }
      >
        Delete
      </Text>
    </View>
    <View
      onLayout={[Function]}
    />
  </View>
  <View
    collapsable={false}
    forwardedRef={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    pointerEvents="auto"
    style={
      Object {
        "transform": Array [
          Object {
            "translateX": -0,
          },
        ],
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#393948",
            "borderRadius": 3,
            "elevation": 5,
            "flexDirection": "row",
            "height": 75,
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.25,
            "shadowRadius": 3.84,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 2,
            "justifyContent": "center",
          }
        }
      >
        <BVLinearGradient
          colors={
            Array [
              4290707360,
              4280600626,
            ]
          }
          endPoint={
            Object {
              "x": 0.2,
              "y": 0.7,
            }
          }
          locations={null}
          startPoint={
            Object {
              "x": 0.8,
              "y": 0,
            }
          }
          style={
            Object {
              "borderRadius": 50,
              "height": 45,
              "width": 45,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "flex": 4,
            "justifyContent": "center",
          }
        }
      >
        <Text
          adjustsFontSizeToFit={true}
          style={
            Object {
              "color": "#FBFBFB",
              "fontFamily": "Exo-Medium",
              "fontSize": 20,
              "marginBottom": 5,
            }
          }
        >
          teste2
        </Text>
        <Text
          adjustsFontSizeToFit={true}
          style={
            Object {
              "color": "#7A7A80",
              "fontFamily": "Exo-Medium",
              "fontSize": 12,
            }
          }
          testID="transactionDate"
        >
          17/08/2020, 4:53:14 pm
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 2,
            "justifyContent": "center",
          }
        }
      >
        <Text
          adjustsFontSizeToFit={true}
          style={
            Object {
              "color": "#24C832",
            }
          }
          testID="transactionValue"
        >
          R$ 3.500,00
        </Text>
      </View>
    </View>
  </View>
</View>
`;
